src/components/EditorView.tsx

```diff
@@ -0,0 +1,329 @@
+  renderSpec,
+  wrapInDecorations,
+  wrapInMarks,
+} from "../nodeViews/render.js";
+
+import { NodeWrapper } from "./NodeWrapper.js";
+import { TextNodeWrapper } from "./TextNodeWrapper.js";
+
+function makeCuts(cuts: number[], node: Node) {
+  const sortedCuts = cuts.sort((a, b) => a - b);
+  const nodes: [Node, ...Node[]] = [node];
+  let curr = 0;
+  for (const cut of sortedCuts) {
+    const lastNode = nodes.pop()!;
+    if (cut - curr !== 0) {
+      nodes.push(lastNode.cut(0, cut - curr));
+    }
+    if (lastNode.nodeSize > cut - curr) {
+      nodes.push(lastNode.cut(cut - curr));
+    }
+    curr = cut;
+  }
+  return nodes;
+}
+
+export type NodeViewComponentProps = {
+  decorations: readonly Decoration[];
+  node: Node;
+  children?: ReactNode | ReactNode[];
+  isSelected: boolean;
+  pos: number;
+};
+
+type EditorStateProps =
+  | {
+      state: EditorState;
+      defaultState?: never;
+    }
+  | {
+      state?: never;
+      defaultState: EditorState;
+    };
+
+export type EditorProps = Omit<
+  DirectEditorProps,
+  "state" | "nodeViews" | "decorations"
+> &
+  EditorStateProps & {
+    keymap?: { [key: string]: Command };
+    nodeViews?: { [nodeType: string]: ComponentType<NodeViewComponentProps> };
+    decorations?: DecorationSet;
+  };
+
+export type Props = EditorProps &
+  DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>;
+
+export function EditorView({
+  children,
+  editable,
+  keymap = {},
+  nodeViews = {},
+  dispatchTransaction: dispatchProp,
+  decorations = DecorationSet.empty,
+  defaultState,
+  state: stateProp,
+  ...mountProps
+}: Props) {
+  const [internalState, setInternalState] = useState<EditorState | null>(
+    defaultState ?? null
+  );
+
+  const posToDOM = useRef(new Map<number, DOMNode>());
+  posToDOM.current = new Map<number, DOMNode>();
+  const domToPos = useRef(new Map<DOMNode, number>());
+  domToPos.current = new Map<DOMNode, number>();
+
+  // We always set internalState above if there's no state prop
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  const state = stateProp ?? internalState!;
+
+  const dispatchTransaction = useMemo(
+    () =>
+      dispatchProp ??
+      ((tr: Transaction) => {
+        setInternalState((prevState) => prevState?.apply(tr) ?? null);
+      }),
+    [dispatchProp]
+  );
+
+  const mountRef = useContentEditable(state, dispatchTransaction);
+
+  useSyncSelection(state, dispatchTransaction, posToDOM, domToPos);
+
+  const onKeyDown: KeyboardEventHandler = (event) => {
+    if (keydownHandler(keymap)(state, dispatchTransaction, event.nativeEvent)) {
+      event.preventDefault();
+    }
+  };
+
+  function buildReactTree(
+    parentElement: JSX.Element,
+    node: Node,
+    pos: number,
+    decorations: DecorationSet
+  ) {
+    const childElements: ReactNode[] = [];
+    if (node.childCount === 0 && node.isTextblock) {
+      childElements.push(<br />);
+    }
+    node.forEach((childNode, offset) => {
```

Comment:
This this is definitely one way to do this, but if we are okay with some intermediate nodes, we may be able to let React memoize levels of this for us. The approach would be to have this be a function component that renders only one level at a time, and wrap it in `React.memo`.

But this isn't identical to how I had been doing things in my earlier attempts to use React with ProseMirror, so I'll keep reading and thinking about the call signature here and whether that works. I mention it so that you can mull it over, too.

Reply:
Yeah; I do think I would rather implement these functions such that they take children as arguments, rather than being responsible for cloning themselves to add children. Was just trying to get the basics down before doing too much cleanup

Reply:
Here's what I had, which was making little React components for each kind of node view, and sharing a render function between them (for some reason).

```
import isShallowEqual from "@wordpress/is-shallow-equal";
import hoistStatics from "hoist-non-react-statics";
import { DOMSerializer } from "prosemirror-model";
import type { Node as ProseMirrorNode } from "prosemirror-model";
import { EditorView } from "prosemirror-view";
import type {
  CustomNodeViewDesc,
  Decoration,
  DecorationSet,
  DirectEditorProps,
  MarkViewDesc,
  NodeViewDesc,
  NodeView as NodeViewSpec,
  ViewDesc,
  WidgetViewDesc,
} from "prosemirror-view";
import { useCallback, useLayoutEffect, useRef, useState } from "react";
import type { ComponentType, ReactNode, Ref, ReactElement } from "react";
import { createPortal } from "react-dom";

function isMarkViewDesc(viewDesc: ViewDesc): viewDesc is MarkViewDesc {
  return "mark" in viewDesc;
}

function isNodeViewDesc(viewDesc: ViewDesc): viewDesc is NodeViewDesc {
  return (
    "node" in viewDesc &&
    "nodeDOM" in viewDesc &&
    "outerDeco" in viewDesc &&
    "innerDeco" in viewDesc
  );
}

function isCustomNodeViewDesc(
  viewDesc: ViewDesc
): viewDesc is CustomNodeViewDesc {
  return isNodeViewDesc(viewDesc) && "spec" in viewDesc;
}

interface ComponentNodeViewSpec extends NodeViewSpec {
  Component: ComponentType<NodeViewDescProps>;
}

interface ComponentNodeViewDesc extends CustomNodeViewDesc {
  spec: ComponentNodeViewSpec;
}

function isComponentNodeViewDesc(
  viewDesc: ViewDesc
): viewDesc is ComponentNodeViewDesc {
  return isCustomNodeViewDesc(viewDesc) && "Component" in viewDesc.spec;
}

function isWidgetViewDesc(viewDesc: ViewDesc): viewDesc is WidgetViewDesc {
  return "widget" in viewDesc;
}

type ViewDescProps<T extends ViewDesc = ViewDesc> = {
  render: (view: EditorView, viewDesc: ViewDesc) => ReactElement | null;
  view: EditorView;
  viewDesc: T;
};

type MarkViewDescProps = ViewDescProps<MarkViewDesc>;

type NodeViewDescProps = ViewDescProps<NodeViewDesc>;

type WidgetViewDescProps = ViewDescProps<WidgetViewDesc>;

function ContentPortal(props: ViewDescProps): ReactElement | null {
  const { render, view, viewDesc } = props;

  const children = viewDesc.children.map((childViewDesc) =>
    render(view, childViewDesc)
  );

  if (viewDesc.contentDOM instanceof Element) {
    return createPortal(children, viewDesc.contentDOM);
  }

  return null;
}

const DocView = (props: NodeViewDescProps) => ContentPortal(props);
DocView.displayName = "Doc";

const MarkView = (props: MarkViewDescProps) => ContentPortal(props);
MarkView.displayName = "Mark";

const NodeView = (props: NodeViewDescProps) => ContentPortal(props);
NodeView.displayName = "Node";

const WidgetView = (props: WidgetViewDescProps) => ContentPortal(props);
WidgetView.displayName = "Widget";

export interface NodeViewProps {
  children?: ReactNode | undefined;
  node: ProseMirrorNode;
  nodeViewRef: Ref<NodeViewSpec>;
  view: EditorView;
  getPos: () => number;
  decorations: Decoration[];
  innerDecorations: DecorationSet;
}

export function nodeView(
  WrappedComponent: ComponentType<NodeViewProps>
): (node: ProseMirrorNode, view: EditorView) => NodeViewSpec {
  const wrappedComponentName =
    WrappedComponent.displayName || WrappedComponent.name || "Component";
  const displayName = `NodeView(${wrappedComponentName})`;

  const Component = (props: NodeViewDescProps) => {
    const { view, viewDesc } = props;
    const { node } = viewDesc;
    const parentViewDesc = viewDesc.parent!;
    const parentDOM = viewDesc.dom?.parentNode!;

    const getPos = useCallback(() => {
      return parentViewDesc.posBeforeChild(viewDesc);
    }, [parentViewDesc, viewDesc]);

    const ref = useCallback(
      (spec: NodeViewSpec | null) => {
        const dom = spec?.dom ?? null;
        if (dom) {
          if (
            dom instanceof HTMLElement &&
            !viewDesc.contentDOM &&
            !dom.hasAttribute("contenteditable")
          ) {
            dom.contentEditable = "false";
          }

          // @ts-expect-error
          dom.pmViewDesc = viewDesc;

          if (viewDesc.dom) {
            parentDOM.replaceChild(dom, viewDesc.dom);
          }
        } else {
          // ProseMirror has already removed the node.
          // Put it back so that React can remove it.
          // :sweat_smile:
          if (viewDesc.dom) {
            parentDOM.insertBefore(viewDesc.dom, parentDOM.firstChild);
          }
        }
        viewDesc.dom = dom;
        viewDesc.nodeDOM = dom;

        const contentDOM = spec?.contentDOM ?? null;
        if (contentDOM) {
          if (viewDesc.contentDOM) {
            Array.from(viewDesc.contentDOM.childNodes).forEach((child) =>
              contentDOM.appendChild(child)
            );
          }
        }
        viewDesc.contentDOM = contentDOM;

        if (isCustomNodeViewDesc(viewDesc)) {
          Object.assign(viewDesc.spec, spec);
        }
      },
      [parentDOM, viewDesc]
    );

    const children = ContentPortal(props);

    return (
      <WrappedComponent
        node={node}
        view={view}
        getPos={getPos}
        decorations={viewDesc.outerDeco}
        innerDecorations={viewDesc.innerDeco}
        nodeViewRef={ref}
        children={children}
      />
    );
  };

  Component.WrappedComponent = WrappedComponent;
  Component.displayName = displayName;
  hoistStatics(WrappedComponent, Component);

  return (node, view) => {
    const document = view.dom.ownerDocument;
    const domSpec = node.type.spec.toDOM?.(node);
    const { dom, contentDOM } = DOMSerializer.renderSpec(document, domSpec!);
    return { dom, contentDOM, Component };
  };
}

const keyCache = new WeakMap<ViewDesc, string>();

function keyCacheGet(viewDesc: ViewDesc): string {
  if (keyCache.has(viewDesc)) {
    return keyCache.get(viewDesc)!;
  } else {
    const key = ((Math.random() * 0xffffff) << 0).toString(16);
    keyCache.set(viewDesc, key);
    return key;
  }
}

function renderViewDesc(
  view: EditorView,
  viewDesc: ViewDesc
): ReactElement | null {
  const key = keyCacheGet(viewDesc);

  if (isComponentNodeViewDesc(viewDesc)) {
    const { Component } = viewDesc.spec;
    return (
      <Component
        render={renderViewDesc}
        view={view}
        viewDesc={viewDesc}
        key={key}
      />
    );
  }

  if (isMarkViewDesc(viewDesc)) {
    return (
      <MarkView
        render={renderViewDesc}
        view={view}
        viewDesc={viewDesc}
        key={key}
      />
    );
  }

  if (isNodeViewDesc(viewDesc)) {
    if (viewDesc.node.isText) {
      return null;
    }

    if (viewDesc.node === view.state.doc) {
      return (
        <DocView
          render={renderViewDesc}
          view={view}
          viewDesc={viewDesc}
          key={key}
        />
      );
    }

    return (
      <NodeView
        render={renderViewDesc}
        view={view}
        viewDesc={viewDesc}
        key={key}
      />
    );
  }

  if (isWidgetViewDesc(viewDesc)) {
    return (
      <WidgetView
        render={renderViewDesc}
        view={view}
        viewDesc={viewDesc}
        key={key}
      />
    );
  }

  return null;
}

export function useEditorView(props: DirectEditorProps) {
  const [mount, setMount] = useState<Node | null>(null);
  const [view, setView] = useState<EditorView | null>(null);
  const hasFocusRef = useRef<boolean>(false);

  useLayoutEffect(() => {
    if (mount && !view) {
      setView(new EditorView({ mount }, props));
    } else if (!mount && view) {
      view.destroy();
      setView(null);
    }
  }, [mount, props, view]);

  useLayoutEffect(() => {
    if (view) {
      // @ts-expect-error
      view.domObserver.start();
      if (hasFocusRef.current) view.focus();
    }
  });

  if (view) {
    const { state: oldState, ...oldProps } = view.props;
    const { state: newState, ...newProps } = props;
    if (isShallowEqual(oldProps, newProps)) {
      if (oldState !== newState) {
        view.updateState(newState);
      }
    } else {
      view.update(props);
    }

    // @ts-expect-error
    view.domObserver.stop();
    hasFocusRef.current = view.hasFocus();
  }

  // @ts-expect-error
  const children = view ? renderViewDesc(view, view.docView) : null;

  return {
    children,
    contentRef: setMount,
    view,
  };
}
```